sycl/doc/design/ESIMDStatelesAccessors.md

```diff
@@ -0,0 +1,165 @@
+stateful intrinsics, and also generate correct annotations. Since it uses the
+same runtime, which relies on parameter annotation when making the wrap/no-wrap
+decision, the runtime part does not need much changes.
+
+## Problem definition
+
+Currently, ESIMD compiler always maps buffer/accessor-based memory accesses to
+stateful accesses, thus imposing the 4Gb datum size limitation on user programs
+with accessors.
+
+## Proposed solution
+
+### Short/mid-term
+
+#### API header changes
+The general idea is to introduce C++ preprocessor macro
+`ESIMD_FORCE_STATELESS_MEM_ACCESS` which will control code generation for the
+stateful memory access APIs - such as:
+
+```cpp
+template <typename Tx, int N, typename AccessorTy,
+          typename Flags = vector_aligned_tag,
+          typename = std::enable_if_t<is_simd_flag_type_v<Flags>>,
+          class T = detail::__raw_t<Tx>>
+__ESIMD_API simd<Tx, N> block_load(AccessorTy acc, uint32_t offset,
+                                   Flags = {});
+
+template <typename T, int N, typename AccessorTy>
+__ESIMD_API std::enable_if_t<(sizeof(T) <= 4) &&
+                                 (N == 1 || N == 8 || N == 16 || N == 32) &&
+                                 !std::is_pointer<AccessorTy>::value,
+                             simd<T, N>>
+gather(AccessorTy acc, simd<uint32_t, N> offsets, uint32_t glob_offset = 0,
+       simd_mask<N> mask = 1);
+```
+
+Implementation of the APIs would follow this pattern:
+
+```cpp
+T stateful_memory_api(accessor acc, uint32 offset, args...) {
+#ifdef ESIMD_FORCE_STATELESS_MEM_ACCESS
+  accessor_elelemt_type *ptr = acc.get_pointer() + offset;
+  return stateless_memory_api(ptr, args...);
+#else
+  <original implementation>
+#endif
+}
+```
+
+The new macro is supposed to be set by users directly or via some logic based on
+other macros set by users.
+
+#### Compiler changes
+
+The API part of the implementation is as simple as above, the compiler
+one is slightly more complicated. Compiler needs to make sure that in presence
+of `ESIMD_FORCE_STATELESS_MEM_ACCESS` macro, the actual memory parameter
+annotation described above is correct and tells that memory is a pointer, not a
+surface index. Parameter annotations are generated by the front-end - these are
+`kernel_arg_accessor_ptr` and `kernel_arg_type` metadata nodes, which are
+then translated to `buffer_t` (for surface) or `svmptr_t` (for pointer)
+metadata annotations consumed by the back-end.
+
+##### Variant 1
+This is the recommended variant. A new driver option is added -
+`-fsycl-esimd-force-stateless-mem-access`. Under this option:
+- SYCL C++ device compiler FE defines the `ESIMD_FORCE_STATELESS_MEM_ACCESS`
+  macro
+- `sycl-post-link` tool is run with a new option
+  `-esimd-force-stateless-mem-access`. Under this option, the tool 
+  configures the LowerESIMD.cpp pass to ignore the `kernel_arg_accessor_ptr`
+  and always generate `svmptr_t` annotation for memory arguments.
+
```

Comment:
Is it possible to mix kernels with stateless accessors and kernels with surface accessors in this variant?

Reply:
In user level code - yes, but compiler will generate stateless for all accessors.